csharp/src/ProtocolBuffers.Test/GeneratedMessageTest.cs

```diff
@@ -510,5 +536,221 @@ public void EmptyPackedValue()
             TestPackedTypes empty = new TestPackedTypes.Builder().Build();
             Assert.AreEqual(0, empty.SerializedSize);
         }
+
+        // oneof tests
+        [Test]
+        public void TestOneofEmumCase()
+        {
+            TestOneof2 message = TestOneof2.CreateBuilder()
+                .SetFooInt(123).SetFooString("foo").SetFooCord("bar").Build();
+            TestUtil.AssertAtMostOneFieldSetOneof(message);
+        }
+
+        [Test]
+        public void TestClearOneof()
+        {
+            TestOneof2.Builder builder = TestOneof2.CreateBuilder().SetFooInt(123);
+            Assert.AreEqual(TestOneof2.FooOneofCase.FooInt, builder.FooCase);
+            builder.ClearFoo();
+            Assert.AreEqual(TestOneof2.FooOneofCase.FooNotSet, builder.FooCase);
+        }
+
+        [Test]
+        public void TestSetOneofClearsOthers()
+        {
+            TestOneof2.Builder builder = TestOneof2.CreateBuilder();
+            TestOneof2 message =
+                builder.SetFooInt(123).SetFooString("foo").BuildPartial();

```

Comment:
Why are we using BuildPartial here? (Just thinking of that going away for proto3, potentially.)